DOCS-11654-http-listener-kt (#1693)
mulesoft/docs-connectors: http/1.6/modules/ROOT/pages/http-listener-ref.adoc

= Configure an HTTP Listener Source - Mule 4
:page-aliases: connectors::http/http-listener-ref.adoc

Anypoint Connector for HTTP (HTTP Connector) *Listener* source enables you to set up an HTTP server and trigger flows when receiving HTTP requests.

> You can choose what methods the source accepts, such as `GET` and `POST` or a list of methods, and in which *Path* to receive HTTP requests, thereby allowing the routing of requests through different flows.

Reviewer: Does the label *Path* have an object that can follow (like *Listener* has "source")?

Reviewer: I wasn't certain how to parse this sentence: whether it was two choices leading to one result (different flows) or two choices leading to two results. It seems like the latter, so optionally consider two sentences:
```suggestion
You can choose what methods the source accepts, such as `GET` and `POST` or a list of methods. You can also choose in which *Path* to receive HTTP requests, thereby allowing requests to be routed through different flows.
```
Is there more to say about why the audience wants to make these choices, or is that obvious?

> When the HTTP *Listener* source receives an HTTP request, the source triggers the flow with the request data. The HTTP body is set as the payload, and the remaining HTTP data as attributes, such as headers and query parameters.

Reviewer: ```suggestion
When the HTTP *Listener* source receives an HTTP request, the source triggers the flow by using the request data. The HTTP body is set as the payload, and the remaining HTTP data is set as attributes, such as headers and query parameters.
```

> When the flow finishes its execution, the HTTP *Listener* source sends an HTTP response based on whether the flow execution was successful or not, which returns different status codes. You can also customize how the source generates HTTP responses in the connector source configuration.

Reviewer: I wonder if the different status codes part is necessary?
```suggestion
When the flow finishes its execution, the HTTP *Listener* source sends an HTTP response based on whether the flow execution was successful or not. You can also customize how the source generates HTTP responses in the connector source configuration.
```
...or maybe
```suggestion
When the flow finishes its execution, the HTTP *Listener* source sends an HTTP status code that indicates whether the flow execution was successful or not. You can also customize how the source generates HTTP responses in the connector source configuration.
```
Maybe provide a reason for customization?
"If you want to <do this different thing from the status codes?>, you can customize the connector source configuration."

== Configure HTTP Listener Connection

> To use an HTTP *Listener* source, you must configure a connection to the HTTP server that listens to the following settings:

Reviewer: Is it accurate to say that a server "listens" to settings?

* *Host* +
Indicates where the requests are sent.
* *Port* +
Indicates where the requests are received.
* *Protocol* +

> Indicates the protocol for communication, it can be HTTP for plain connections or HTTPS for TLS secure connections.

Reviewer: ```suggestion
Indicates the protocol for communication, either HTTP for plain connections or HTTPS for TLS secure connections.
```

To enable secure connections through HTTPS, you must define a TLS context in the connection and provide:

* A keystore for the server
* A truststore if you need two-way authentication

Refer to the xref:mule-runtime::tls-configuration.adoc[TLS Configuration] documentation for details.

> Additionally, you can configure a *Base path* that applies to all sources using the configuration.

Reviewer: Similar to the comment on Line 6, does *Base path* have an object? "value"?

The following example shows how to configure the HTTP *Listener* connection in Studio:

> . In the *Mule Palette* view, select *HTTP>Listener*.

Reviewer: Are there no spaces around ">" in the UI label?

[start=4]
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP *Listener* in the app.
. Set *Host* to `0.0.0.0` and *Port* to `8081`.

image::http-listener-2.png[HTTP Listener global configuration window with Host field set to 0.0.0.0 and Port field to 8081]

[start=6]
. Click the *TLS* tab.
. Set *TLS configuration* to *Edit inline*.
. In the *Trust Store Configuration* section, set *Path* to `keystore.jks` and *Password* to `MyP455W0rD`.
. Click *OK*.

image::http-listener-3.png[Trust Store Configuration window with Path field set to keystore.jks and Password field to MyP455W0rD]

In the Configuration XML editor, the `host`, `port`, `path` and `trust-store` configurations look like these:

[source,xml,linenums]
----
<http:listener-config name="HTTP_Listener_config" d >
		<http:listener-connection host="0.0.0.0" port="8081" >
			<tls:context >
				<tls:trust-store path="keystore.jks" password="MyP455W0rD" />
			</tls:context>
		</http:listener-connection>
	</http:listener-config>
  <flow name="test" >
		<http:listenerconfig-ref="HTTP_Listener_config" path="/path"/>
	</flow>
----

> To configure the *Base path* for the source:

Reviewer: Ditto comment Line 30; GLOBAL for these value names/labels.

. In the HTTP *Listener* configuration screen, click the plus sign (*+*) next to the *Connector configuration* field to configure a global element.
. In *HTTP Listener config*, set *Base path* to `api/v1`.
. Click *OK*.

image::http-listener-4.png[HTTP Listener global configuration window with Base path field set to api/v1]

In the Configuration XML editor, the `basePath` configuration looks like this:

[source,xml,linenums]
----
<http:listener-config name="HTTP_Listener_config" basePath="api/v1">
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>
----

[NOTE]
If you deploy a project to CloudHub, set *Host* to *0.0.0.0* so the connection listens to all the interfaces of the machine. +

> If you deploy locally for testing purposes, set *Host* to *localhost* to receive only requests that are generated locally. The app is not vulnerable to external threats.

Reviewer: The first sentence advises to test locally.
The second sentence does not seem to apply to the reason to test locally but to be a separate issue (security)?

> You can also configure advanced settings that can define whether the connections are persistent and if they are not, the timeout the connections will have.

Reviewer: ```suggestion
You can also configure advanced settings that can define whether the connections are persistent and, if they are not, the timeout for the nonpersistent connections.
```

== Configure Paths

Because an HTTP *Listener* source acts as an event trigger within flows, it  must reference the configuration to use and the *Path* where the source listens for HTTP requests.

> Paths can be static, which requires exact matches, or feature placeholders. Placeholders can be:

Reviewer: ```suggestion
Paths can be either static, which requires exact matches, or feature placeholders. Placeholders can be:
```

* Wildcards (`*`) +

> Match against anything they are compared to.

Reviewer: ```suggestion
Match anything they are compared to.
```

* Parameters (`{param}`) +

> Match not only against anything but also capture those values on a URI parameters map.

Reviewer: ```suggestion
Match not only anything they are compared to but also capture those values on a URI parameters map.
```

> [NOTE]

Reviewer: Consider moving this to be part of Line 101?

Wildcards at the end of a path can help provide better error messages by capturing requests to unmanaged resources.

The following examples show different *Path* configurations for three HTTP *Listener* sources that have a *Base path* set to `api/v1`:

* `account/mulesoft/main-contact` +
Matches only the exact path request `+http://awesome-company.com/api/v1/account/mulesoft/main-contact+`.
* `account/{accountId}/main-contact` +

> Matches all path requests structured similarly, such as `+http://awesome-company.com/api/v1/account/salesforce/main-contact+`, and save `salesforce` as the value of `accountId`.

Reviewer: "...matches...and saves...."
```suggestion
Matches all path requests structured similarly, such as `+http://awesome-company.com/api/v1/account/salesforce/main-contact+`, and saves `salesforce` as the value of `accountId`.
```

* `account/{accountId}/*` +

> Matches all path requests different from `main-contact`, such as `+http://awesome-company.com/api/v1/account/mulesoft/users+`, and save `mulesoft` as the value of `accountId`.

Reviewer: ditto re. "saves" ^

When the Mule app has multiple HTTP *Listener* sources, HTTP requests are always routed to the most specific source. In the previous examples, the first HTTP *Listener* source receives a request with `accountId: mulesoft` and suffix `main-contact`, and the second source receives any different `accountId` value.

== Configure Allowed Methods

> HTTP requests are routed based on the HTTP method received. By default, an HTTP *Listener* source supports all methods. You can also restrict the methods of your choice, even configure custom ones.

Reviewer: To me, "you can also" sounds like "in addition to," and I think you mean an alternative?
```suggestion
HTTP requests are routed based on the HTTP method received. By default, an HTTP *Listener* source supports all methods, but you can restrict the available methods or even configure custom ones.
```
...and tell WHY a user might want or need to do this?

The following example shows how to configure *Allowed methods* in Studio:

> . In the *Mule Palette* view, select *HTTP>Listener*.

Reviewer: same ? as earlier, about na space around >...?
GLOBAL; I'll stop asking

. Drag *Listener* to the Studio canvas.
. In the HTTP *Listener* source configuration screen, click the *Advanced* tab.
. Set *Allowed methods* to `GET`.

image::http-listener-5.png[HTTP Listener configuration window and Allowed methods field set to GET]

In the Configuration XML editor, the `allowedMethods` configuration looks like this:
[source,xml,linenums]
----
<http:listener path="/test" allowedMethods="GET" config-ref="HTTP_Listener_config"/>
----

> When a Mule app has multiple HTTP *Listener* sources defined, requests are routed to the first source matching the method, so default sources should always be defined last.

Reviewer: More active?
```suggestion
When a Mule app defines multiple HTTP *Listener* sources, requests are routed to the first source matching the method, so default sources should always be defined last.
```
...based on the assumption that provided sources are preferred to the default. Is this clear to the audience?

> The following example shows different HTTP *Listener* sources configuration where the *Allowed method* is set to `GET`, `POST` and `PUT`. Therefore, HTTP requests are routed to different flows based on the restriction of data visibility. You might allow certain users to see your data but only a few users to modify it:

Reviewer: ```suggestion
The following example shows different HTTP *Listener* source configurations, in which the *Allowed method* is set to `GET`, `POST` and `PUT`. These different configurations cause HTTP requests to be routed to different flows, enabling you to specify and restrict user access to your data.
```

</flow>
----

== Configure Response Streaming Mode

> When HTTP Connector manages response bodies, the connector considers the type of data to send and uses _chunked_ encoding when the body size is not clear. This behavior is also known as streams with no size information.

Reviewer: Maybe this is not technically accurate, but when you introduce a term, like chunking, and then say that it is "also known as," what follows is typically a synonymous term, not a description. So...something like this?
```suggestion
When HTTP Connector manages response bodies, the connector considers the type of data to send and uses _chunked_ encoding when the body size is not clear. This behavior is also known as _response streaming mode_.
```

Reviewer: HOWEVER, looking at the following text, I think that this might be closer to what you mean:
```suggestion
When HTTP Connector manages response bodies, the connector considers the type of data to send and uses _chunked_ encoding when the body size is not clear. This causes streaming unlimited by data size.
```
...which results in <some negative result>?

> To change this behavior configure the *Response streaming mode* to any of these options:

Reviewer: ```suggestion
To control this behavior, configure *Response streaming mode* to any of these options:
```
The options that follow control both defined and not defined size scenarios, so "change" is probably not the right word, unless you want to limit what follows to only those options that disallow chunking.

* *AUTO (default)* +
Uses `Content-Length` encoding if a size is defined for the body, otherwise uses `Transfer-Encoding: chunked`.
* *ALWAYS* +

> Uses `Transfer-Encoding: chunked` regardless of any size data present.

Reviewer: "regardless of whether body size is defined or not"?

* *NEVER* +
Uses `Content-Length` encoding, consuming streams if necessary to determine the data size.

> The following example shows how to configure  *Response streaming mode* in Studio. In the example, the main contact data for an account always returns `Content-Length` encoding:

Reviewer: Is this still what you mean?
```suggestion
The following example shows how to use Studio to configure *Response streaming mode* to always use `Content-Length` encoding:
```

. In the *Mule Palette* view, select *HTTP>Listener*.
. Drag *Listener* to the Studio canvas.
. Set *Path* to `account/{accountId}/main-contact`.
. In the HTTP *Listener* source configuration screen, click the *Advanced* tab.
. Set *Allowed methods* to `GET`.
. Set *Response streaming mode* to `NEVER`.

image::http-listener-6.png[HTTP Listener configuration window with Allowed methods field set to GET and Response streaming mode field set to NEVER]

> In the Configuration XML editor, the `allowedMethods` and `responseStreamingMode` configurations look like these:

Reviewer: ```suggestion
In the Configuration XML editor, the `allowedMethods` and `responseStreamingMode` configurations look like this:
```

[source,xml,linenums]
----
<flow name="main-contact-read">
  <http:listener path="account/{accountId}/main-contact" allowedMethods="GET" responseStreamingMode="NEVER" config-ref="HTTP_Listener_config"/>
  <!-- fetch main contact for accountId -->
</flow>
----

> 

Reviewer: BREAK here to commit.

== Configure Read Timeout

To configure the read timeout for the source, in the HTTP *Listener* configuration window, set the *Read timeout* field to the desired value:

image::http-listener-7.png[HTTP Listener configuration window with Read timeout field set to 30000]

> In the Configuration XML editor, the `readTimeout` configuration looks like these:

Reviewer: ```suggestion
In the Configuration XML editor, the `readTimeout` configuration looks like this:
```
the configuration (one thing) looks like this (single example)

[source,xml,linenums]
----
<http:listener-config >
		<http:listener-connection host="0.0.0.0" port="8081" readTimeout="3000" />
	</http:listener-config>
----

== Configure Reject Invalid Transfer Encoding

> To reject requests with invalid Transfer-Encoding headers, in the HTTP *Listener* configuration window, select the *Reject invalid transfer encoding* field:

Reviewer: Select a field? Do you mean select a checkbox, or type a value in a field, or select a value from a menu in a field?

image::http-listener-8.png[HTTP Listener configuration window with Reject invalid transfer encoding field selected]

> In the Configuration XML editor, the `rejectInvalidTransferEncoding` configuration looks like these:

Reviewer: ```suggestion
In the Configuration XML editor, the `rejectInvalidTransferEncoding` configuration looks like this:
```

[source,xml,linenums]
----
<http:listener-config rejectInvalidTransferEncoding="true">
		<http:listener-connection host="0.0.0.0" port="8081" readTimeout="3000" />
</http:listener-config>
----

== Listening to HTTP Requests

> When an HTTP *Listener* source receives an HTTP request, the data from the HTTP request line includes the method, request path, query, URI parameters, and headers as attributes. The body sets the payload, and the `Content-Type` header sets the  MIME type, which enables other components to inspect the payload MIME type. For example, DataWeave works with an HTTP payload without requiring any input information. Additionally, if an HTTP request contains an `X-Correlation-ID` or `MULE_CORRELATION_ID` (for interoperability with Mule 3) header, it sets the message's correlation ID for traceability.

Reviewer: ```suggestion
When an HTTP *Listener* source receives an HTTP request, the data from the HTTP request line includes the method, request path, query, URI parameters, and headers as attributes. The body sets the payload, and the `Content-Type` header sets the  MIME type, which enables other components to inspect the payload MIME type. For example, DataWeave works with an HTTP payload without requiring any input information. Additionally, if an HTTP request contains an `X-Correlation-ID` or `MULE_CORRELATION_ID` header (for interoperability with Mule 3), it sets the message's correlation ID for traceability.
```
Please clarify to what "it" refers: request? header?

}
----

The attributes include:

* method: `POST`
* listenerPath: `api/v1/account/{accountId}/main-contact`
* requestPath: `api/v1/account/salesforce/main-contact`
* relativePath: `account/salesforce/main-contact`

> * queryParams: a multi-map with entries `overwrite=true`, `notify=jane.doe` and `notify=admin`

Reviewer: ```suggestion
* queryParams: a multimap with entries `overwrite=true`, `notify=jane.doe`, and `notify=admin`
```
GLOBALLY change "multi-map" to "multimap," per online search and our style ref.

* uriParams: a map with entry `accountId => salesforce`

> * headers: a multi-map with entries `Host=localhost:8081`, `Content-Type=application/json`, `Content-Length=166` and `X-Correlation-ID=9cf32672-4f0b-4e8b-b988-40c13aae85b4`

Reviewer: Keep an eye out for series commas:
```suggestion
* headers: a multi-map with entries `Host=localhost:8081`, `Content-Type=application/json`, `Content-Length=166`, and `X-Correlation-ID=9cf32672-4f0b-4e8b-b988-40c13aae85b4`
```

A DataWeave expression such as `#[payload.name ++ ' ' ++ payload.surname]` returns `John Doe` because DataWeave correctly interprets the JSON data.

> A DataWeave expression such as `#['Received a ' ++ attributes.method ++ ' request for account ' ++ attributes.uriParams.accountId ++ '. The following users are notified: ' ++ (attributes.queryParams.*notify joinBy ', ')]` returns `Received a POST request for account salesforce. The following users are notified: admin, jane.doe`.

Reviewer: The first sentence is not a complete sentence.
The next two sentences do not require colons, because the first "lists" only one user and the second "lists" only two. Reword the first to "XYZ is notified" and also why that is followed by the "returns" phrase? Then reword the second to "XYZ and ABC are notified."

Reviewer: I did not understand Lines 279-281 without context; you might check with a writing peer for this product line.

[NOTE]

> Multi-maps are similar to maps except they allow several values for a given key. Multi-maps return the first value when using a single-value selector (`.`), but they allow to retrieve all values when using the multiple-value selector (`.*`).

Reviewer: Better grammatically, but please assure technical fidelity:
```suggestion
_Multimaps_ are maps that allow one key to have several values. A multimap returns only the first of several values when invoked using a single-value selector (`.`), but returns all values when invoked using the multiple-value selector (`.*`).
```
I was not sure about "invoked": are maps "called" or "configured" or "enabled" or...?

=== HTTP Request Multipart Form-Data Example

The following example shows an HTTP request of an HTML form received by an HTTP *Listener* source. Additionally, the example shows how to use DataWeave expressions to read `multipart/form-data` content.

Based on the following HTML form:

> [source,xml,linenums]

Reviewer: Avoid breaking a sentence around an example, leaving the audience waiting for the rest. 
For example, fully introduce the example and then provide it, or introduce the example and then follow with a sentence that explains or adds to it.

--489691234097965980223899--
----

> You can use DataWeave expressions to access and read each HTTP request parameter data through the `parts` object, either by name or item number. For example, to access the second part of the HTTP request that contains the `Corporate Logo` parameters data, you can use `payload.parts.logo` or `payload.parts[1]`. The latter is useful when a name is not provided.

Reviewer: "through" = "by using"?
"...data by using either a name or item number `parts` object."  ?

Reviewer: If you want to access "the second part" you use `payload.parts[1]`? "second" and "1" seem inconsistent.

Reviewer: Instead of "The latter," consider "Using the item number".

Reviewer: My best guess:
```suggestion
You can use DataWeave expressions to access and read each HTTP request parameter data via the `parts` object plus a name or, in the absence of a name, an item number. For example, to access the second part of the HTTP request that contains the `Corporate Logo` parameters data, you can use `payload.parts.logo` or `payload.parts[2]`. 
```

Within each part, you can access its content and headers. For example, `payload.parts.color.content` returns `blue`, while `payload.parts.logo.headers.'Content-Type'` returns `application/png`.

You can also access the filename of a part. For example, the `Content-Disposition` header is parsed to allow an expression like `payload.parts.logo.headers.'Content-Disposition'.filename`, which in this case returns `MuleSoft_logo.png`.

Refer to the xref:dataweave::dataweave-formats.adoc[Formats Supported by DataWeave] documentation to learn about reading and writing multipart content.

== Sending HTTP Responses

> After the triggered flow finishes its execution, the HTTP *Listener* source sends an HTTP response with the result. If the flow executes successfully, the source sends an HTTP regular response. Otherwise, the source sends an HTTP error response.

Reviewer: Does this mean the same thing?
```suggestion
After the triggered flow finishes its execution, the HTTP *Listener* source sends either an HTTP regular response, if execution is successful, or an HTTP error response.
```

Reviewer: OR, including the following para, here is another option that might increase readability:
```suggestion
After the triggered flow finishes its execution, the HTTP *Listener* source sends either of the following default results:
* Successful execution: a 200 status code and the message payload as the body
* Unsuccessful execution: a 500 status code with the flow error's description as the body
```

> By default, an HTTP regular response returns a 200 status code and the message payload as the body, while an HTTP error response returns a 500 status code with the flow error's description as the body.

Reviewer: See suggestion for combining this with prev.

In the HTTP *Listener* source configuration, you can customize HTTP responses by providing the following parameters. Note that you can use DataWeave to generate content for each parameter, and variables to propagate data from the flow:

* *Body*
* *Headers*
* *Status code*
* *Reason phrase*

=== Create Sending HTTP Responses Mule App Example

The following example shows how to configure HTTP responses for the HTTP *Listener* source in Studio. The example is a Mule app flow where an endpoint stores logos for an account:

> * When the storage is successful, an HTTP regular response returns with status code `201`, reason phrase `Created`, and body as `Corporate Logo has been stored as a MuleSoft logo`.

Reviewer: ```suggestion
* When storage is successful, an HTTP regular response returns the status code `201`, the reason phrase `Created`, and the body `Corporate Logo has been stored as a MuleSoft logo`.
```

> * When the storage fails, an HTTP error response returns where the status code is defined through the `errorCode` variable (if available) or a `500` code by default.

Reviewer: ```suggestion
* When storage fails, an HTTP error response returns a status code defined through the `errorCode` variable (if available) or `500` by default.
```

** The custom header returns `X-Time`.
** The body returns `Corporate Logo could not be stored`.

> ** If there is a `CONNECTIVITY` error storing the logo, the returned status code is `504`, while any other errors result in a `500` status code.

Reviewer: Can these lines be made parallel with the components of the description of success?

Reviewer: Maybe we need a breakdown of 
* Success
** Status
** Reason
** Body
* Failure
** Status
** Reason
** Body

> While a reason phrase is not defined for error responses, HTTP Connector attempts to define a reason phrase based on the status code. Thereby, a `Gateway Timeout` or `Internal Server Error` error can return for the previous explained HTTP responses scenarios.

Reviewer: See suggestion above.

To create the Mule app in Studio, follow these steps:

. In the *Mule Palette* view, select *HTTP>Listener*.
. Drag *Listener* to the Studio canvas.
. Set *Path* to `/account/{accountId}/logo`.
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP *Listener* in the app.
. Set *Host* to `0.0.0.0` and *Port* to `8081`.
. Click *OK*.
. In the HTTP *Listener* configuration screen, click the *Responses* tab.
. In the *Response* section, set these values:
+
* *Body* : `output text/plain --- vars.logoName ++ ' has been stored as a ' ++ vars.accountId ++ ' logo.'`
* *Status code* : `201`
* *Reason phrase* : `Created`

[start=9]
. In the *Error Response* section, set these values:
+
* *Body* : `vars.logoName ++ ' could not be stored.'`
* *Headers* :
+

> [source,DataWeave,linenums]

Reviewer: I wonder if it would be more easily understood to show the task first and the response explanations after?

----
output application/java
        ---
        {
          "X-Time" : "50s" // 4
        }
----
* *Status code* : `vars.errorCode default 500`.

image::http-listener-9.png[HTTP Listener Responses configuration window with Body, Header, Status Code and Reason phrase fields]

[start=10]
. Drag a *Set Variable* component to the right of HTTP *Listener*.
. Set *Name* to `logoName`.
. Set *Value* to `payload.parts.name.content`.
. Drag another *Set Variable* component to the right of first *Set Variable* component.
. Set *Name* to `accountId`.
. Set *Value* to `attributes.uriParams.accountId`.

> . Drag an *Insert*  operation to the right of  the second *Set Variable* component.

Reviewer: ```suggestion
. Drag an *Insert* operation to the right of the second *Set Variable* component.
```
